src/main/java/fi/vm/yti/datamodel/api/v2/service/SchemaService.java

```diff
@@ -1,161 +1,264 @@
+		propertyResource.addProperty(model.getProperty(SH.node.getURI()), model.createResource(targetShape));
+
 		return propertyResource;
-		
-	}		
-	
+	}
 
-	/**
+	private void handleDatatypeProperty(String propID, Entry<String, JsonNode> entry, Model model, String schemaPID,
+			Resource nodeShapeResource, boolean isRequired, boolean isArrayItem) {
+		Resource propertyResource = addDatatypeProperty(propID + "/" + entry.getKey(), entry.getValue(), model,
+				schemaPID, entry.getValue().get("type").asText());
+		nodeShapeResource.addProperty(model.getProperty(SH.property.getURI()), propertyResource);
+		if (!isArrayItem) {
+			handleRequiredProperty(entry.getValue(), model, propertyResource, isRequired);
+		} 
+		if (entry.getValue().get("type").asText().equals("string") & entry.getValue().has("pattern")) {
+			propertyResource.addProperty(model.getProperty(SH.pattern.getURI()), entry.getValue().get("pattern").asText());
+		}
+	}
 	
-	Handles an object property and creates the corresponding SHACL (Node)Shape.
+	private String capitaliseNodeIdentifier(String propID) {
+		List<String> propIDSplit = new LinkedList<String>(Arrays.asList(propID.split("/")));
+		String partToCapitalise = propIDSplit.get(propIDSplit.size() - 1);
+		return ((String.join("/", propIDSplit)) + "/" + partToCapitalise.substring(0, 1).toUpperCase() + partToCapitalise.substring(1));
+	}
 	
-	@param propID The property ID.
-	@param node The JSON node containing the property details.
-	@param schemaPID The schema PID.
-	@param model The RDF model.
-	*/
+	/**
+	 * 
+	 * Handles an object property and creates the corresponding SHACL (Node)Shape.
+	 * 
+	 * @param propID    The property ID.
+	 * @param node      The JSON node containing the property details.
+	 * @param schemaPID The schema PID.
+	 * @param model     The RDF model.
+	 */
 	private void handleObject(String propID, JsonNode node, String schemaPID, Model model) {
 
-		Resource nodeShapeResource = model.createResource(schemaPID + "#" + propID);
+		String propIDCapitalised = capitaliseNodeIdentifier(propID);
+		String nameProperty = propID.substring(propID.lastIndexOf("/") + 1);
+		Resource nodeShapeResource = model.createResource(schemaPID + "#" + propIDCapitalised);
 		
-		nodeShapeResource.addProperty(RDF.type, model.getResource(SHACL.NodeShape.getURI()));
-		nodeShapeResource.addProperty(MSCR.localName, propID);
-		if(node.has("description"))
+		nodeShapeResource.addProperty(RDF.type, model.getResource(SH.NodeShape.getURI()));
+		nodeShapeResource.addProperty(MSCR.localName, nameProperty);
+		nodeShapeResource.addProperty(model.getProperty(SH.name.getURI()), nameProperty);
+		
+		if (node.has("description"))
 			nodeShapeResource.addProperty(DCTerms.description, node.get("description").asText());
-		nodeShapeResource.addProperty(model.getProperty(SHACL.name.getURI()), propID);
-		if(node == null || node.get("properties") == null) {
+		if (node == null || node.get("properties") == null) 
```

Comment:
If you are checking for node == null that should really happen before the node.has("description") call, because if the node is null the current code will thrown an exception before the execution gets to this row.